src/jit/gentree.cpp

```diff
@@ -12142,6 +12255,185 @@ GenTreePtr Compiler::gtFoldExprCompare(GenTreePtr tree)
+
+                    // These handle constants should be class handles.
+                    assert(op1ClassFromHandle->IsIconHandle(GTF_ICON_CLASS_HDL));
+                    assert(op2ClassFromHandle->IsIconHandle(GTF_ICON_CLASS_HDL));
+                }
+            }
+        }
+
+        GenTree* compare = gtNewOperNode(oper, TYP_INT, op1ClassFromHandle, op2ClassFromHandle);
+
+        // Drop any now-irrelvant flags
+        compare->gtFlags |= tree->gtFlags & (GTF_RELOP_JMP_USED | GTF_RELOP_QMARK | GTF_DONT_CSE);
+
+        return compare;
+    }
+
+    // Just one operand creates a type from a handle.
+    //
+    // If the other operand is fetching the type from an object,
+    // we can sometimes optimize the type compare into a simpler
+    // method table comparison.
+    //
+    // TODO: if other operand is null...
+    if (op1Kind != TPK_GetType && op2Kind != TPK_GetType)
+    {
+        return tree;
+    }
+
+    GenTree* const opHandle = op1IsFromHandle ? op1 : op2;
+    GenTree* const opOther  = op1IsFromHandle ? op2 : op1;
+
+    // Tunnel through the handle operand to get at the class handle involved.
+    GenTree* const opHandleArgument = opHandle->gtCall.gtCallArgs->gtOp.gtOp1;
+    GenTree*       opHandleLiteral  = opHandleArgument;
+
+    // Unwrap any GT_NOP node used to prevent constant folding
+    if ((opHandleLiteral->gtOper == GT_NOP) && (opHandleLiteral->gtType == TYP_I_IMPL))
+    {
+        opHandleLiteral = opHandleLiteral->gtOp.gtOp1;
+    }
+
+    // In the ngen case, we have to go thru an indirection to get the right handle.
+    if (opHandleLiteral->gtOper == GT_IND)
+    {
+        // Handle indirs should be marked as nonfaulting.
+        assert((opHandleLiteral->gtFlags & GTF_IND_NONFAULTING) != 0);
+        opHandleLiteral = opHandleLiteral->gtOp.gtOp1;
+    }
+
+    // If, after tunneling,  we don't have a constant handle, bail.
+    if ((opHandleLiteral->gtOper != GT_CNS_INT) || (opHandleLiteral->gtType != TYP_I_IMPL))
+    {
+        return tree;
+    }
+
+    // We should have a class handle.
+    assert(opHandleLiteral->IsIconHandle(GTF_ICON_CLASS_HDL));
+
+    // Fetch the compile time handle, and use it to ask the VM if
+    // this kind of type can be equality tested by a simple method
+    // table comparison.
+    CORINFO_CLASS_HANDLE clsHnd = CORINFO_CLASS_HANDLE(opHandleLiteral->gtIntCon.gtCompileTimeHandle);
+
+    if (!info.compCompHnd->canInlineTypeCheckWithObjectVTable(clsHnd))
+    {
+        return tree;
+    }
+
+    // We're good to go.
+    JITDUMP("Optimizing compare of obj.GetType()"
+            " and type-from-handle to compare method table pointer\n");
+
+    // opHandleArgument is the method table we're looking for.
+    GenTree* const knownMT = opHandleArgument;
+
+#ifdef LEGACY_BACKEND
+    // Fetch object method table from the object itself
+    GenTree* const objMT = gtNewOperNode(GT_IND, TYP_I_IMPL, opOther->gtCall.gtCallObjp);
```

Comment:
I am surprised to see an indef LEGACY_BACKEND here. I would have expected that we didn't do the optimization with legacy backend.

Reply:
What makes the LEGACY_BACKEND path different here anyway?

Reply:
This is a holdover from how the code was before, when it was in morph. 

Not sure how things got to the state they were just before my changes, but here's a guess.

Some time in the past, calls and intrinsics were done differently for the legacy backend and this optimization was supported (or if not, somebody went to a lot of trouble to make it look like it was).

Then at some point, somebody unified the way calls and intriniscs were handled, but did not update this optimization. So the pattern matches that guard this optimization stopped matching for the legacy backend, and the transformation part specific to the legacy backend went stale.

Now that the importer paths are unified it would be simplest to just remove the ifdef'd code and have this entire optimization be common code.

But I can also just disable this opt for the legacy backend.